compiler/qsc_partial_eval/src/lib.rs

```diff
@@ -1104,6 +1106,78 @@ impl<'a> PartialEvaluator<'a> {
         Ok(EvalControlFlow::Continue(Value::Tuple(values.into())))
     }
 
+    fn eval_expr_unary(
+        &mut self,
+        un_op: UnOp,
+        value_expr_id: ExprId,
+        unary_expr_span: Span, // For diagnostic purposes only.
+    ) -> Result<EvalControlFlow, Error> {
+        let value_expr = self.get_expr(value_expr_id);
+        let value_control_flow = self.try_eval_expr(value_expr_id)?;
+        let EvalControlFlow::Continue(value) = value_control_flow else {
+            return Err(Error::Unexpected(
+                "embedded return in unary operation expression".to_string(),
+                value_expr.span,
+            ));
+        };
+
+        // Get the variable type corresponding to the value the unary operator acts upon.
+        let Some(eval_variable_type) = try_get_eval_var_type(&value) else {
+            return Err(Error::Unexpected(
+                format!("invalid type for unary operation value: {value}"),
+                value_expr.span,
+            ));
+        };
+
+        // The leading positive operator is a no-op.
+        if matches!(un_op, UnOp::Pos) {
+            let control_flow = EvalControlFlow::Continue(value);
+            return Ok(control_flow);
+        }
+
+        // For all the other supported unary operations we have to generate an instruction, so create a variable to
+        // store the result.
+        let variable_id = self.resource_manager.next_var();
+        let rir_variable_type = map_eval_var_type_to_rir_type(eval_variable_type);
+        let rir_variable = rir::Variable {
+            variable_id,
+            ty: rir_variable_type,
+        };
+
+        // Generate the instruction depending on the unary operator.
+        let value_operand = map_eval_value_to_rir_operand(&value);
+        let instruction = match un_op {
+            UnOp::Neg => {
+                let constant = match rir_variable_type {
+                    rir::Ty::Integer => Operand::Literal(Literal::Integer(-1)),
+                    rir::Ty::Double => Operand::Literal(Literal::Double(-1.0)),
```

Comment:
turns out there is a floating point negation operation in LLVM, though that probably gets inserted by optimization. We should consider whether we want to use `fneg` and add it into RIR or keep this as a multiplication for simplicity. See [fneg](https://llvm.org/docs/LangRef.html#fneg-instruction) in the LLVM docs.

Relatedly, this made me notice that we are always emitting the QIR with text for LLVM integer operations, and technically those need to be the floating point alternatives if we expect these to work properly, ie: `mul` for integer operands and `fmul` for double operands. @idavis FYI, so we can decide if we should support this in QIR codegen or treat it as an error for now.

Reply:
Looking again at the tests, we actually aren't testing this functionality right now. The best thing might be treating a negation of a dynamic double as a failure (maybe `Error::Unimplemented`) for now and punting on putting these instructions into the RIR.

Reply:
I was thinking about supporting doubles in partial evaluation even if they are currently not supported in RIR to QIR. The reason being that it's almost the same amount of work to support them, and we also would rely on a verification pass after partial evaluation is done that would report an error if we produced something that is not supported by the intended target so it wouldn't go to the RIR to QIR stage.